### PR TITLE
[CLEANUP beta] Remove array controller

### PR DESCRIPTION
**Pull request by cibernox**
NO MERGE: Depends on #11479

Probably there is a few more tasks to do, but in different PRs:
- [x] Remove Ember.SortableMixin? Is private and it's not used anymore internally. Or alternatively make it public.
- [x] prefix with underscore `beforeObserver/addBeforeObserver/removeBeforeObserver/beforeObserversFor` and deprecate pubic interfaces in the `Ember` namespace

As a curiosity, ember is finally under 500KB again:
499KB

### CODE REVIEW

**Comment by stefanpenner** [comment]
> Remove Ember.SortableMixin? Is private and it's not used anymore internally. Or alternatively make it public.

yes this should also be extracted.

Is there an addon for arraycontroller already?

**Comment by cibernox** [comment]
Extracted addon for ObjectController and ArrayController: https://github.com/cibernox/ember-proxy-controllers

**Comment by stefanpenner** [comment]
* this should also remove sortable mixin
* this should move `Ember.beforeObserver` -> `Ember._privateBeforeObserver`

**Comment by cibernox** [comment]
This weekend I'll extract this to the addon and finish this.

**Comment by stefanpenner** [comment]
After some discussion with @tomdale regarding itemController. We suspect, itemController hooks should be left in the view layer, but only become functional when the legacy ArrayController addon is present.

File: packages/ember-htmlbars/lib/keywords/each.js [hunk 1]
```diff
@@ -3,18 +3,10 @@
...
 */
 
-import ArrayController from 'ember-runtime/controllers/array_controller';
-
 export default function each(morph, env, scope, params, hash, template, inverse, visitor) {
   let getValue = env.hooks.getValue;
-  let firstParam = params[0] && getValue(params[0]);
   let keyword = hash['-legacy-keyword'] && getValue(hash['-legacy-keyword']);
 
-  if (firstParam && firstParam instanceof ArrayController) {
```
Comment (stefanpenner) [inline packages/ember-htmlbars/lib/keywords/each.js]: I believe these hooks will need to stay (as deprecated) for some reasonable period of 2.x.

Obviously requiring the `ArrayController` for the `instanceof` wont work, so we will need to revert to duckTyping of some kind. Maybe something like: `firstParam.isArrayController`  ?

File: packages/ember-views/lib/views/legacy_each_view.js [hunk 2]
```diff
@@ -14,34 +13,7 @@ export default View.extend(EmptyViewSupport, {
   template: legacyEachTemplate,
   tagName: '',
 
-  _arrayController: computed(function() {
```
Comment (stefanpenner) [inline packages/ember-views/lib/views/legacy_each_view.js]: this may also need to stay, or become part of the legacy addons code-base.

**Comment by stefanpenner** [comment]
> Remove generateController abstraction?

This may need to want until routableComponents land. Although `generateController` should only generate plan controllers, not object or array.

**Comment by stefanpenner** [comment]
> Remove ControllerMixin?

this stays around, since Ember.Controller is still needed as we did not land RoutableComponents.

**Comment by stefanpenner** [comment]
@cibernox thanks for all your work here, I am super excited for this to land!

File: packages/ember-views/lib/views/legacy_each_view.js (hunk 2 above)
Comment (cibernox) [inline packages/ember-views/lib/views/legacy_each_view.js]: https://github.com/cibernox/ember-proxy-controllers/blob/master/app/initializers/ember-legacy-controllers.js#L11-L34

File: packages/ember-htmlbars/lib/keywords/each.js (hunk 1 above)
Comment (cibernox) [inline packages/ember-htmlbars/lib/keywords/each.js]: added back

File: packages/ember-views/lib/views/legacy_each_view.js (hunk 2 above)
Comment (stefanpenner) [inline packages/ember-views/lib/views/legacy_each_view.js]: nice!

**Comment by stefanpenner** [comment]
@cibernox this looks good (except for the failing tests) Im available tomorrow and thursday to help you in anyway to get this in.

**Comment by stefanpenner** [comment]
now needs a rebase

lets be sure we get the addon to production readiness shortly.

**Comment by tomdale** [comment]
Is @cibernox working on extracting the addon as well?

**Comment by stefanpenner** [comment]
Yes. He has it on his account   Appears to be complete (to the degree discussed )

**Comment by cibernox** [comment]
Apparently the tests with the old jquery are consistently failing in travis, but locally they work. Is is a known issue?

**Comment by stefanpenner** [comment]
> Is is a known issue?

No, master appears to be green.

**Comment by cibernox** [comment]
Ok, it was random. It passed after 5 tries.

**Comment by stefanpenner** [comment]
sorry the second commit should likely also be BETA

@rwjblue c/d?

**Comment by rwjblue** [comment]
Thank you!

**Pull request merged**